### ISSUE TITLE
Add "destructureInSignature: never" to destructuring-assignment

### ISSUE THREAD

**Issue by max-prtsr** [jsx-eslint/eslint-plugin-react#3330]
That would forbid to use

```tsx
const MyComponent = ({ name }: {name: string}) => <h1>Hello, { name } </h1>
```

Reason: 
- in TS, it looks less readable
- forces our developers to use only one variant 
- seeing `props` in code (e.g., in review) helps identify react components with 100% sure

**Comment by max-prtsr** [comment]
> I don’t think it’s less readable in TS; and usually the props types are defined outside the component in a type/interface.

Maybe,  but an option of using props and destructuring in the function body would be nice for me :)

**Comment by ljharb** [comment]
Destructuring in the signature has concrete benefits, whether using TS or not - it avoids the props object being accessible at all, which avoids a huge class of bugs caused by mutating or exposing that mutable props object.

This isn't a mere aesthetic difference, and I don't think it would be a good idea to have a rule that encouraged it.

**Comment by max-prtsr** [comment]
> Destructuring in the signature has concrete benefits, whether using TS or not - it avoids the props object being accessible at all, which avoids a huge class of bugs caused by mutating or exposing that mutable props object.

Nobody uses `props.x`, because we're using restructuring on the next line

> This isn't a mere aesthetic difference, and I don't think it would be a good idea to have a rule that encouraged it.

There is

We have components with 10-20 props and putting them in a function definition is less readable for me.

Looks like basically, you're saying "we don't use it - nobody needs it", which is untrue.

**Comment by ljharb** [comment]
That should suggest to you that "having 10-20 props" is what causes the readability problem, not "where they're destructured".

Either way, I think this:
```jsx
function Foo({
  a,
  b,
  c,
  d,
  e,
  …
  z,
}) {
```
is pretty objectively more readable than this:
```jsx
function Foo(props) {
  const {
    a,
    b,
    c,
    d,
    e,
    …
    z,
  } = props;
}) {
```
_especially_ with long lists, because it duplicates the `props` reference, and the longer the list, the farther apart they are.

What I'm saying is that not all stylistic preferences are equal, and sometimes, a choice actively harms the maintainability of a codebase, and in these cases, it's better for an eslint plugin (which by definition are meant to be opinionated) not to tacitly endorse the pattern by providing configuration for it.

You are welcome to modify the rule in your own plugin, and use that.

**Comment by ljharb** [comment]
If WebStorm doesn't work well with an idiomatic style, WebStorm is broken, not the style - file a bug with them.

Obviously if you want to access the props object then it's fine to move it to the body, but passing props objects to children wholesale is an antipattern.

**Comment by ezze** [comment]
@ljharb Thanks for the answer. Sad to hear that you won't support it. The rule could be more flexible and give the ability to adapt to different code style. That's what ESLint was created for, I guess. Probably we have to implement it by ourselves if there are no any better alternatives.

**Comment by ljharb** [comment]
When it's purely style, yes! However, never destructuring in the signature actually makes your code worse and less robust, which will cause bugs in your code - so it's not purely a matter of style, and it would be reckless for me to enable that.

**Comment by ezze** [comment]
I prefer to destructure it in the first line of function body, it doesn't make my code worse and less robust obviously. There is [prefer-destructuring](https://eslint.org/docs/latest/rules/prefer-destructuring) rule allowing to avoid not destructured properties.

And what if I want to combine `props` with some other variables using spread operator and pass the result to child component? It's very useful when child component has common properties with its parent (wrapper component).

The notice about multiline destructuring in signature is still actual.

**Comment by ljharb** [comment]
Then you should be explicitly reconstructing what you need. You shouldn't be able to reference the original props object at all, lest it get mutated.

**Comment by devandtravel** [comment]
@ljharb Just to be clear about the spread in the function signature won't protect you from bugs.

What a spread operator does? It deep copies the data if it is not nested. For nested data, it deeply copies the topmost data and shallow copies of the nested data. Destructuring in the signature just protects me from reference the original props object but allows me to reference nested data of props object. This is only a partial protection in case there is a nested data in the props.

It is entirely the responsibility of the developer to protect props from mutations. You may still need to be careful when modifying nested objects or arrays to avoid unintentionally modifying the original object.

While it's true that using the spread operator and destructuring can help protect your code from unintended mutations, not using destructuring in the function signature does not necessarily make your code worse or less robust.

In conclusion, while using the spread operator and destructuring in the function signature can provide some benefits, it's up to the developer to choose their preferred coding style and ensure they are handling data carefully to prevent unintended mutations.

**Comment by ljharb** [comment]
The bugs I'm concerned about are exclusively about mutating the props object, since that's the one React looks at when rerendering.

**Comment by devandtravel** [comment]
Of course, in the context of the React discussion, this makes sense.

**Comment by max-prtsr** [comment]
> wever, never destructuring in the signature actually makes your code worse and less robust, which will cause bugs in your code - so it's not purely a matter of style, and it would be reckless for me to enable that.

It's so untrue...

**Comment by RomainLanz** [comment]
So what is the status of this issue? Why is it closed?

It seems fair to ask for a `never` option for the `destructureInSignature`. Why would we be able to force the destructuring in the signature but not force the destructuring to **not be** in the signature?

Please reconsider this issue as it seems to be a nice-to-have feature and should not break anything.

**Comment by ljharb** [comment]
@RomainLanz have you read the entire thread? not having it in the signature can cause bugs, so it's not purely a style choice.

**Comment by RomainLanz** [comment]
Hey @ljharb! 👋🏻 

I've thoroughly read the entire thread once again to make sure I didn't miss any points

From what I gathered, it seems there's a concern that some people might accidentally mutate the props object without fully understanding its implications. While this is a valid concern, I believe it's more of a potential misunderstanding rather than an actual bug.

Is there any other potential "bug" that you think I might have missed?

**Comment by ljharb** [comment]
Nope, that's the thing. My extensive experience tells me that this is indeed an actual bug, and at times a catastrophic one, since mutating props invalidates any predictions about React rendering.

**Comment by max-prtsr** [comment]
@ljharb you're basically selling your opinion as only valid.

**Comment by ljharb** [comment]
@max-prtsr nope, just the one that makes decisions *here*. an eslint rule is supposed to be opinionated - choices are for style, but any choice that causes bugs is indeed one that folks should be actively prevented from making.

**Comment by RomainLanz** [comment]
> Nope, that's the thing. My extensive experience tells me that this is indeed an actual bug, and at times a catastrophic one, since mutating props invalidates any predictions about React rendering.

I completely understand your concern. Mutating the props object can indeed lead to unwanted behavior and introduce bugs in an application.

However, I'd like to mention that the [`no-param-reassign`](https://eslint.org/docs/latest/rules/no-param-reassign) ESLint rule was specifically created to prevent parameter mutation. You can leverage the props option of this rule to ensure that any attempt to mutate a parameter will be caught and prevented, like this:

```ts
function Component(props) {
    props.foo = "value";  // Error
}
```

Additionally, when you're destructuring your params inside the signature, it can be a bit challenging to assign default values. In such cases, it's much easier to do so when performing the destructuring outside, like this:

```ts
function Component(props) {
    const { foo = 'bar' } = props;
}
```

Another advantage of destructuring outside the signature is that you cannot mark the props 'as const' when doing it inside, which helps prevent accidental reassignments:

```ts
function Component({ foo }) {
    foo = 'acme'  // Works, which might lead to bugs
}
```

```ts
function Component(props) {
    const { foo } = props;
    
    foo = 'acme'  // Error, making it safer
}
```

In the end, both approaches have their pros and cons, and it's essential to choose the one that best suits your specific use case. This is why I believe it's valuable to support and allow both approaches within the ESLint rules, as they cater to different coding styles and scenarios.

**Comment by connorjs** [comment]
> The bugs I'm concerned about are exclusively about mutating the props object, since that's the one React looks at when rerendering.
>
> -- <https://github.com/jsx-eslint/eslint-plugin-react/issues/3330#issuecomment-1499700065>

For what it’s worth, React only looks at props for re-rendering in the memo case, not by default. And if you’re memo-ing, you (a) have to consider non-shallow props (which the current rule does NOT help) and (b) should already have clear understanding/goal/tests that lead you to memo.

---

I found this issue also hoping for `destructureInSignature: never` _because_ the rule allows `never` at the root. I liked @RomainLanz’s examples and wanted to share my own. (I also admit I will just eslint-disable for now.)

Thus, the following is an error, which surprises me. While the `...baseProps` may be questionable, it is common and _correct_ logic for wrapping components.

```tsx
export function Button(props: ButtonProps) {
	// Destructure non-HTML button props. Also sets defaults.
	const {
		children,
		icon,
		onClick,
		variant = `secondary`,
		...baseProps
	} = props;
	
	return null; // Would render
}
```

The main reason I prefer the `never` in the signature (but generally prefer destructuring) is for inline TypeScript types that I see many developers prefer for helper components (helper methods). Destructing also _is_ a good practice like you mentioned, but there are TS style concerns that are 100% subjective (there’s no objective answer on the pure style side, which I think does exist).

```tsx
export function Foo({bar, baz}: {bar: string, baz: number}) {
  const num = baz * 2; // or whatever logic
  return `Hello, ${bar} (${num})`;
}
```

```tsx
export function Foo(props: {bar: string, baz: number}) {
	const num = props.baz * 2; // or whatever logic	
	return `Hello, ${props.bar} (${num})`;
}
```

---

All of that being said, this is why we have `// eslint-disable-next-line`, which works fine for the wrapping component (majority) use case.

```tsx
// Destructure non-HTML button props. Also sets defaults.
const { blah } = props;
```

becomes

```tsx
// eslint-disable-next-line react/destructuring-assignment -- Destructure non-HTML button props. Also sets defaults.
const { blah } = props;
```

_I wrote up the full post in case my React understanding is wrong. Please do correct me._
